### DISCUSSION TITLE
write_registers() throws "unpack requires a buffer of 4 bytes"

### DISCUSSION THREAD

**Discussion by Frankstar** [pymodbus-dev/pymodbus#2279] [Help] [answered]
- Python: 3.12.4
- OS: Fedora 39
- Pymodbus: 3.6.9 and 3.7.0
- Modbus Hardware: Solarman Stick Logger LSE-3 (Modbus TCP)

### sample code
```
from pymodbus.client import ModbusTcpClient
client = ModbusTcpClient(host='10.0.100.34', port=8899)
try:
    response = client.write_registers(0x1110, 0, slave=1)
except Exception as e:
    print(f"General exception: {e}")
finally:
    client.close()
```

Im trying to write a single register (with write_registers since the Hardware needs it that way).
The example code is throwing the error `unpack requires a buffer of 4 bytes`.
The interesting thing is that the register gets set correctly.
So something might be wrong with the response.
reading via `read_holding_registers()` is working fine.

3.6.9 throwed:
```
unpack requires a buffer of 4 bytes
Modbus IO exception Modbus Error: [Input/Output] Unable to decode request
Error writing to register 4368: Modbus Error: [Input/Output] Unable to decode request
```

3.7.0 is throwing:
```
unpack requires a buffer of 4 bytes
```

### more simplified Code:
```
from pymodbus.client import ModbusTcpClient
client = ModbusTcpClient(host='10.0.100.34', port=8899)
response = client.write_registers(0x1110, 0, slave=1)
print(response)
client.close()
```

Traceback:
```
(most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/client/mixin.py", line 334, in write_registers
    return self.execute(
           ^^^^^^^^^^^^^
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/client/base.py", line 237, in execute
    return self.transaction.execute(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/transaction.py", line 267, in execute
    self.client.framer.processIncomingPacket(
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/framer/old_framer_base.py", line 154, in processIncomingPacket
    self.frameProcessIncomingPacket(single, callback, slave, tid=tid)
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/framer/old_framer_socket.py", line 87, in frameProcessIncomingPacket
    if (result := self.decoder.decode(data)) is None:
                  ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/factory.py", line 236, in decode
    return self._helper(message)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/factory.py", line 266, in _helper
    response.decode(data[1:])
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/pdu/register_write_message.py", line 264, in decode
    self.address, self.count = struct.unpack(">HH", data)
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^
struct.error: unpack requires a buffer of 4 bytes
>>> print(response)
```

### output with debug logging:

```
DEBUG:pymodbus.logging:Connection to Modbus server established. Socket ('10.0.100.132', 49226)
DEBUG:root:Sending value: 0 to register: 4368
DEBUG:pymodbus.logging:[0]
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 2
DEBUG:pymodbus.logging:SEND: 0x0 0x2 0x0 0x0 0x0 0x9 0x1 0x10 0x11 0x10 0x0 0x1 0x2 0x0 0x0
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x0 0x2 0x0 0x0 0x0 0x14 0x1 0x10 0x11 0x10 0x0 0x1 0x5 0x30 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0
DEBUG:pymodbus.logging:Got response!!!
DEBUG:pymodbus.logging:Processing: 0x0 0x2 0x0 0x0 0x0 0x14 0x1 0x10 0x11 0x10 0x0 0x1 0x5 0x30 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0
DEBUG:pymodbus.logging:Factory Response[WriteMultipleRegistersResponse': 16]
ERROR:root:General exception: unpack requires a buffer of 4 bytes
Traceback (most recent call last):
  File "<stdin>", line 8, in write
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/client/mixin.py", line 334, in write_registers
    return self.execute(
           ^^^^^^^^^^^^^
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/client/base.py", line 237, in execute
    return self.transaction.execute(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/transaction.py", line 267, in execute
    self.client.framer.processIncomingPacket(
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/framer/old_framer_base.py", line 154, in processIncomingPacket
    self.frameProcessIncomingPacket(single, callback, slave, tid=tid)
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/framer/old_framer_socket.py", line 87, in frameProcessIncomingPacket
    if (result := self.decoder.decode(data)) is None:
                  ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/factory.py", line 236, in decode
    return self._helper(message)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/factory.py", line 266, in _helper
    response.decode(data[1:])
  File "/home/username/.local/lib/python3.12/site-packages/pymodbus/pdu/register_write_message.py", line 264, in decode
    self.address, self.count = struct.unpack(">HH", data)
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^
struct.error: unpack requires a buffer of 4 bytes
>>> 
```

Any Ideas?
I cant see the "real" error but i'm not sure if its a bug/issue of pymodbus.

**Comment by janiversen** [comment]
Your call to write_registers is not correct, please read the documentation or at least follow the parameter definitions.

**Reply by Frankstar** [reply]
sorry jan - but whats wrong? im even more confused now
maybe i can't see the forest for the trees...

documentation states:

> write_registers(address: int, values: list[int] | int, slave: int = 1, skip_encode: bool = False)→ T[]
> Parameters:
> address – Start address to write to
> values – List of values to write, or a single value to write
> slave – (optional) Modbus slave ID
> skip_encode – (optional) 

I could pass the values as a list (values: list[int] | int)
tried that with `client.write_registers(0x1110, [0], slave=1)` -> same error. 

Than i tried with `skip_encode=True` and passed my value formatted with `encoded_value = struct.pack('>H', 0)` -> same error.
What else i'm missing?

**Reply by Frankstar** [reply]
ok than i'm misunderstanding the documentation.
I read it as "you can pass a list or just an int value" or even a list with 1 value.

i know about write_register but:
a) the device is not allowing function code 0x06
b) it would be cleaner for me personally just using one function for all.
c) it "works" with `write_registers` but somehow the response code is broken for pymodbus

anyway -> sry for wasting your time than and thanks for helping!
maybe i will find a solution for this problem.

**Comment by janiversen** [comment]
I am just controlling your different statements:

- The example code of 3.7.0 works perfectly (client_async_calls), and sends the request correctly, and gets a correct response.
- I added a couple of extra tests e.g. [10] as value, and it works nicely.
- The parameter description was wrong, that's corrected.

Please make a proper debug log (pymodbus_apply_debug) with 3.7.0, I think your problem might be that your device responds wrongly.

**Reply by janiversen** [reply]
just checked your traceback, it fails in "self.decoder.decode", that is when the response is received, so now it is very interesting to see what you actually receive.

**Reply by janiversen** [reply]
Just to avoid misunderstandings, I would like to see the test with 3.7.0, since I cannot make bug finding in 3.6.9

**Reply by Frankstar** [reply]
the traceback was with 3.7.0
3.6.9 just throwed 2 more errors earlier that day - thats why i was mentioning it.
Thought it might help understanding my problem.

**Comment by janiversen** [comment]
The response:
```
DEBUG:pymodbus.logging:RECV: 0x0 0x2 0x0 0x0 0x0 0x14 0x1 0x10 0x11 0x10 0x0 0x1 0x5 0x30 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0
DEBUG:pymodbus.logging:Got response!!!
```
Is by far too large ! The standard defines 7 bytes header + 5 bytes response.

**Reply by Frankstar** [reply]
ok that's interesting. learned something new! thank you.
so there is no real way of fixing this i guess.

is there a way to suppress/ignore the response?

---
just if anyone will find this discussion.
the device is a Sofar HYD Inverter.
The documentation:
![image](https://github.com/user-attachments/assets/fd0a74d7-831e-4582-8e11-92b4588e5bfc)

**Reply by Frankstar** [reply]
since i'm pretty new to the modbus protocol i asked my "friend" chatGPT about the response...
TL;DR: you are right (never doubted it!) - the response is odd.

>The response byte sequence `0x0 0x2 0x0 0x0 0x0 0x14 0x1 0x10 0x11 0x10 0x0 0x1 0x5 0x30 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0` represents:
>
>Transaction ID: 0x0002
>Protocol ID: 0x0000 (Modbus)
>Length: 0x0014 (20 bytes)
>Unit ID: 0x01
>Function Code: 0x10 (Write Multiple Registers)
>Starting Address: 0x1110
>Quantity of Registers: 0x0001 (1 register)
>Byte Count: 0x05 (5 bytes of data)
>
>The byte count of 0x05 and the subsequent register data (0x30 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0) might suggest some additional information about the state or a confirmation of what was written, although for a typical write multiple registers response, you usually don’t see this much data. It’s possible that the additional bytes are providing some kind of extended status or extra data beyond the standard response.

**Reply by janiversen** [reply]
Well that is a serial description, NOT tcp/ip !!! you might have a rs-485 - tcp converter installed, but then you probably should use the rtu framer.

Apart from that the request/response are correct.

**Reply by janiversen** [reply]
There are no byte count in the response (see your own attachment).

**Reply by Frankstar** [reply]
i also tried communicate with the device directly via rs485 - but got some weird timeouts.
Sofar Support suggest than the LSE-3 usb to ethernet adapter and stated "This one is talking ModBus TCP natively over the port 8899."

anyway - you helped alot.
Thanks!

**Reply by Frankstar** [reply]
> There are no byte count in the response (see your own attachment).

yes - and still i got some data ....
but i couldn't have troubleshooted this far without you. thanks for your time!
Now i know its no pymodbus problem - i dont know how i can workaround it - but thats another story.

**Reply by Frankstar** [reply, marked as answer]
> Is by far too large ! The standard defines 7 bytes header + 5 bytes response.

just to clarify it -> the wrong (too much) response data is than leading to
```
struct.error: unpack requires a buffer of 4 bytes)
```
?
(so i can mark this question/discussion as answered )
